### PR TITLE
refactor: remove global network from transaction controller 

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Upgrade `@metamask/transaction-controller` to remove all usages of the global network.

Specifically:

- Create global network utils including:
  - `getGlobalChainId`
  - `getGlobalNetworkClientId`
  - `getGlobalEthQuery`
- Add `networkClientId` to calls to:
  - `addTransaction`
  - `estimateGas`
  - `getNonceLock`
  - `startIncomingTransactionPolling`
  - `updateIncomingTransactions`
- Update calls to:
  - `getTransactions`
  - `wipeTransactions`
- Add `networkClientId` to Redux transaction object.
- Add `networkClientId` to `Stake` type in `StakeContext`.
- Add `networkClientId` argument to transaction utils:
  - `getMethodData`
  - `isSmartContractAddress`
- Add `networkClientId` to test data.

## **Related issues**

Fixes: [#3499](https://github.com/MetaMask/MetaMask-planning/issues/3499)

## **Manual testing steps**

Full regression of all transaction flows.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by sleepytanya** [comment]
Bitrise QA builds https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/256a579f-cde8-4d78-a063-4c08f56d1da8?tab=workflows

**Review by pedronfigueiredo — DISMISSED** [review]

**Comment by sleepytanya** [comment]
@matthewwalsh0 the `undefined is not an object (evaluating n.chainId)` is also present in RC 7.36.0


https://github.com/user-attachments/assets/c3d3c792-0380-4e08-a01d-b7a94084d229

**Comment by sleepytanya** [comment]
`Main` build https://app.bitrise.io/app/be69d4368ee7e86d/installable-artifacts/b2bf9484b6bdf580

Nonce error is not present:

Screenshot 2024-11-22 at 10 45 52

File: app/components/Views/NetworkSelector/NetworkSelector.tsx [hunk 1]
```diff
@@ -381,7 +381,7 @@ const NetworkSelector = () => {
       AccountTrackerController.refresh();
 
       setTimeout(async () => {
-        await updateIncomingTransactions();
+        await updateIncomingTransactions([clientId]);
```
Comment (EtherWizard33) [inline app/components/Views/NetworkSelector/NetworkSelector.tsx:384]: Thanks @matthewwalsh0 
i see the upcoming transaction now takes a clientId, I guess that is a chain id? or is the clientId something like mobile vs extension?

Comment (matthewwalsh0) [reply app/components/Views/NetworkSelector/NetworkSelector.tsx:384]: This is a `networkClientId` so identifies a specific network client / provider from the `NetworkController`.

**Review by Matt561 — DISMISSED** [review]
Staking changes have been tested successfully locally. 👍

**Review by infiniteflower — DISMISSED** [review]
Tested swaps and working with STX and non STX

**Review by wachunei — APPROVED** [review]

**Review by EtherWizard33 — APPROVED** [review]
LGTM, but I’m unsure of all implications. If there are breaking changes, @adonesky1 should review, as this may be a prerequisite for his per-dapp selected network work.  

This PR and a sync with Alex could be good topics for the next weekly meeting. It might be worth holding off on merging until then to ensure alignment, as it wasn’t discussed yesterday.

**Review by bergeron — APPROVED** [review]

**Comment by matthewwalsh0** [comment]
> LGTM, but I’m unsure of all implications. If there are breaking changes, @adonesky1 should review, as this may be a prerequisite for his per-dapp selected network work.
> 
> This PR and a sync with Alex could be good topics for the next weekly meeting. It might be worth holding off on merging until then to ensure alignment, as it wasn’t discussed yesterday.

There are no breaking changes for the user, this is ultimately just a controller upgrade PR with the minimum changes to support a transaction controller that no longer relies on the global network.

This is indeed required to support per-dapp selected network and has been planned and discussed in multiple multi-chain syncs 👍 

We don't want to delay this any further as it's blocking further adoption of the transaction controller versions in mobile.

**Review by tommasini — APPROVED** [review]
LGTM! We can address the circular dept after the PR be merged, since this touches multiple ownership

File: app/util/networks/global-network.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import EthQuery from '@metamask/eth-query';
+import {
+  NetworkClientId,
+  NetworkController,
+} from '@metamask/network-controller';
+import { Hex } from '@metamask/utils';
+import Engine from '../../core/Engine';
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply app/util/networks/global-network.ts:7]: I also prefer to isolate new logic and not continue to grow `Engine.ts`.

The reason the utils need to reference Engine are so every caller doesn't need to access the global controllers directly, but we can encapsulate that knowledge in here.

Hopefully this becomes a non-issue as EVM multichain should ultimately remove these selectors and any reliance on the global network in the background.

**Review by vinistevam — APPROVED** [review]

File: app/util/networks/global-network.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tommasini) [reply app/util/networks/global-network.ts:7]: Sounds great!

**Review by christopherferreira9 — APPROVED** [review]
LGTM

**Review by Matt561 — APPROVED** [review]

**Review by infiniteflower — APPROVED** [review]

**Pull request merged**